src/org/rascalmpl/util/maven/SimpleResolver.java

```diff
@@ -90,9 +99,83 @@ private Path calculateJarPath(String groupId, String artifactId, String version,
         return pomLocation.resolveSibling(fileName + ".jar");
     }
 
+    // Note: Because we download both the metadata and the pom files from the first repo that
+    // has the file available, we could end up in a situation where the metadata file comes from 
+    // a different repo than the pom file. This should not pose a problem as pom files with the
+    // same version should be identical. If they are not, there is something seriously wrong with
+    // the versioning in one of the repos and we are helpless to fix that anyway.
+    private Metadata downloadArtifactMetadata(String groupId, String artifactId, String versionSpec)
+        throws UnresolvableModelException {
+
+        // TODO: deal with repository filters etc in settings.xml
+        String metadataUrl = String.format("/%s/%s/maven-metadata.xml", groupId.replace('.', '/'), artifactId);
+
+        MetadataXpp3Reader reader = new MetadataXpp3Reader();
+        for (var repoDownloader : availableRepostories) {
+            if (repoDownloader.getRepo().getLayout().equals("legacy")) {
+                // TODO: support legacy repo
+                continue;
+            }
+
+            String metadataSource = repoDownloader.read(metadataUrl);
+            if (metadataSource != null) {
+                try {
+                    return reader.read(new StringReader(metadataSource));
+                }
+                catch (IOException e) {
+                    throw new UnresolvableModelException("IOException trying to parse metadata", groupId, artifactId, versionSpec, e);
+                }
+                catch (XmlPullParserException e) {
+                    throw new UnresolvableModelException("Xml exception trying to parse metadata", groupId, artifactId, versionSpec, e);
+                }
+            }
+        }
+        throw new UnresolvableModelException("Could not download artifact metadata from available repositories", groupId,
+            artifactId, versionSpec);
+    }
+
+    private String findLatestMatchingVersion(Metadata metadata, String groupId, String artifactId, String versionRange) throws UnresolvableModelException {
+        try {
+            VersionRange range = VersionRange.createFromVersionSpec(versionRange);
+            Versioning versioning = metadata.getVersioning();
+            String latest = null;
+            DefaultArtifactVersion latestVersion = null;
+
+            for (String currentVersion : versioning.getVersions()) {
+                DefaultArtifactVersion candidate = new DefaultArtifactVersion(currentVersion);
+                if (range.containsVersion(candidate)) {
+                    if (latestVersion == null || candidate.compareTo(latestVersion) > 0) {
+                        latestVersion = candidate;
```

Comment:
this code reads a bit strange since `latestVersion` is not a string but `latest` is. While `currentVersion` is a string, and `candidate` (instead of `current`?) is not a string. These 4 relates variables are inconsitantly named.

Also can we not keep only the `DefaultArtifactVersion`  around and at the end extract the string representation back from it?